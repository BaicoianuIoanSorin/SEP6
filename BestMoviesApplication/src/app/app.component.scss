.left-container {
  display: flex;
  align-items: center;
}

.right-container {
  display: flex;
  align-items: center;
  justify-content: flex-end;
  margin-left: auto;
}

.left-container button,
.right-container button {
  margin-right: 10px; /* Adjust the margin as per your preference */
}
