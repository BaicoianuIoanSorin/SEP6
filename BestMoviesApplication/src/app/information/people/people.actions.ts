export class PeopleFetchInfo {
  static readonly type = 'People - Fetching info';
  constructor() {
  }
}

export class PeopleReset {
  static readonly type = 'People - Reset';
  constructor() {
  }
}

export class PeopleSearchName {
  static readonly type = 'People - Search by person name';
  constructor(public personName: string) {
  }
}

export class PeopleSearchReset {
  static readonly type = 'People - Search Reset';
  constructor() {
  }
}
