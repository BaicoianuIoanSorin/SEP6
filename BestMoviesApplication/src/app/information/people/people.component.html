<ng-container *ngIf="isFetching$ | async; then dataFetching; else dataFetched">
</ng-container>

<ng-template #dataFetched>
  <div class="people-information">
    <nb-card>
      <nb-card-body>

        <nb-tabset listenWindowScroll>
          <nb-tab tabTitle="Stars"
                  [badgeText]="starsSize$ | async"
                  badgeStatus="success">
            <div class="stars">
              <nb-card>
                <nb-card-header class="header-information">
          <span class="title">
            Stars
          </span>

                  <div class="right-side">
                    <bm-search
                      (searchEvent)="onSearch(getPeopleType().STAR,$event)"
                      (resetSearch)="resetSearch(getPeopleType().STAR)"
                      [isFiltered]="starsAreFiltered$ | async"
                      [placeHolder]="starsPlaceholder"
                    >

                    </bm-search>
                  </div>
                </nb-card-header>
                <nb-card-body>
                  <nb-list
                    nbInfiniteList
                    [threshold]="500"
                    (bottomThreshold)="loadNext(getPeopleType().STAR)"
                    class="movie-container"
                    *ngIf="stars$ | async as stars"
                  >
                    <nb-list-item *ngFor="let star of stars">
                      <nb-card size="medium" class="poster-people">
                        <nb-card-header class="header-poster-people">
                          <div class="people-name">{{ star.name }}</div>
                        </nb-card-header>
                        <nb-card-body class="body-poster-people">
                          <div class="show-more-button">
                            <button
                              nbButton
                              status="primary"
                              (click)="redirectToOverviewPageFor(getPeopleType().STAR,star.id)"
                            >
                              <nb-icon icon="film-outline"></nb-icon>
                              Show more
                            </button>
                          </div>
                          <img *ngIf="star.profilePicture"
                               [src]="star.profilePicture" [alt]="star.name">
                          <img *ngIf="!star.profilePicture"
                               src="../../../assets/no-poster-available.jpg" [alt]="star.name">
                        </nb-card-body>
                        <nb-card-footer class="footer-poster-people">
                          <div>
                            <span class="person-birth">Year of birth: {{ star.birthYear }}</span>
                          </div>
                        </nb-card-footer>
                      </nb-card>
                    </nb-list-item>
                  </nb-list>
                </nb-card-body>
              </nb-card>
            </div>
          </nb-tab>
          <nb-tab tabTitle="Directors"
                  [badgeText]="directorsSize$ | async"
                  badgeStatus="warning">
            <div class="directors">
              <nb-card>
                <nb-card-header class="header-information">
          <span class="title">
            Directors
          </span>

                  <div class="right-side">
                    <bm-search
                      (searchEvent)="onSearch(getPeopleType().DIRECTOR,$event)"
                      (resetSearch)="resetSearch(getPeopleType().DIRECTOR)"
                      [isFiltered]="directorsAreFiltered$ | async"
                      [placeHolder]="directorsPlaceholder"
                    >

                    </bm-search>
                  </div>
                </nb-card-header>
                <nb-card-body>
                  <nb-list
                    nbInfiniteList
                    [threshold]="500"
                    (bottomThreshold)="loadNext(getPeopleType().DIRECTOR)"
                    class="movie-container"
                    *ngIf="directors$ | async as directors"
                  >
                    <nb-list-item *ngFor="let director of directors">
                      <nb-card size="medium" class="poster-people">
                        <nb-card-header class="header-poster-people">
                          <div class="people-name">{{ director }}</div>
                        </nb-card-header>
                        <nb-card-body class="body-poster-people">
                          <div class="show-more-button">
                            <button
                              nbButton
                              status="primary"
                              (click)="redirectToOverviewPageFor(getPeopleType().DIRECTOR, director.id)"
                            >
                              <nb-icon icon="film-outline"></nb-icon>
                              Show more
                            </button>
                          </div>
                          <img *ngIf="director.profilePicture"
                               [src]="director.profilePicture" [alt]="director.name">
                          <img *ngIf="!director.profilePicture"
                               src="../../../assets/no-poster-available.jpg" [alt]="director.name">
                        </nb-card-body>
                        <nb-card-footer class="footer-poster-people">
                          <div>
                            <span class="person-birth">Year of birth: {{ director.name }}</span>
                          </div>
                        </nb-card-footer>
                      </nb-card>
                    </nb-list-item>
                  </nb-list>
                </nb-card-body>
              </nb-card>
            </div>
          </nb-tab>
        </nb-tabset>
      </nb-card-body>
    </nb-card>
  </div>
</ng-template>

<ng-template #dataFetching>
  <nb-spinner
    status="basic"
    class="spinner">
  </nb-spinner>
</ng-template>
