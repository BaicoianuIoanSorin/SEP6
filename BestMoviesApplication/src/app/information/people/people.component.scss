.people-information {

  .header-information {
    display: flex;
    justify-content: space-between;
  }

  .movie-container {
    display: grid;
    grid-template-columns: repeat(auto-fill, minmax(200px, 1fr));
    grid-gap: 1rem;
    max-width: 1200px;
    margin: 0 auto;

    .poster-people {
      position: relative;
      width: 100%;
      border-radius: 15px 15px 15px 15px;
      box-shadow: 8px 6px 20px 2px #6d6d6d;

      .header-poster-people {
        .people-name {
          height: 2.8em;
        }
      }

      .body-poster-people {
        padding: unset;
        overflow: hidden;
        position: relative;

        .show-more-button {
          position: absolute;
          top: 50%;
          left: 50%;
          transform: translate(-50%, -50%);
          opacity: 0;
          transition: opacity 0.3s ease;
          display: flex;
          align-items: center;
          justify-content: center;
          z-index: 1; /* Ensure the button is above the image */
        }

        img {
          width: 100%;
          height: 100%;
          object-fit: cover;
          object-position: center;
        }
      }

      .footer-poster-people {
        margin-top: 0px;

        .person-birth {
          margin-top: 1rem;
        }
      }

    }

    .poster-people:hover {
      opacity: 0.6;
      cursor: pointer;
    }

    .poster-people:hover .body-poster-people .show-more-button {
      opacity: 1;
    }
  }

  .stars {
    width: 100%;
  }

  .directors {
    width: 100%;

    nb-card-body {
      display: grid;
      gap: 0.4rem;
    }
  }
}

.spinner {
  color: nb-theme(color-primary-500);
}
