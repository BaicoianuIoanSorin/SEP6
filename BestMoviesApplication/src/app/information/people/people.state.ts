import {Action, State, StateContext} from "@ngxs/store";
import {Injectable} from "@angular/core";
import {Person} from "src/model/person";
import {Movie} from "src/model/movie";
import {NbToastrService} from "@nebular/theme";
import {
  MoviesFetchInfo,
  MoviesReset,
  MoviesSearchReset,
  MoviesSearchTitle
} from "src/app/information/movies/movies.actions";
import {produce} from "immer";
import {moviesMock} from "src/util/mocks/movies_mock";
import {MoviesStateModel} from "src/app/information/movies/movies.state";
import {
  PeopleFetchInfo,
  PeopleReset,
  PeopleSearchName,
  PeopleSearchReset
} from "src/app/information/people/people.actions";
import {peopleMock} from "src/util/mocks/people_mock";
import {Star} from "src/model/star";
import {StarsService} from "src/api/stars/stars.service";
import {DirectorService} from "src/api/directors/director.service";

export interface PeopleStateModel {
  isFetching: boolean;
  isFiltered: boolean;
  stars: Star[];
  directors: Star[];
}

export const defaultsState: PeopleStateModel = {
  isFetching: false,
  isFiltered: false,
  stars: [],
  directors: [],
}

@State<PeopleStateModel>( {
  name: 'peoplePage',
  defaults: defaultsState,
})

@Injectable()
export class PeopleState {
  constructor(
    private toastrService: NbToastrService,
    private starsService: StarsService,
    private directorsService: DirectorService,
  ) {
  }

  @Action(PeopleFetchInfo)
  async peopleFetchInfo(
    { getState, setState }: StateContext<PeopleStateModel>) {

    let currentState = getState();

    let newState = produce(currentState, draft => {
      draft.isFetching = true;
    })

    setState(newState);
    currentState = newState;

    let stars: Star[] = [];
    let directors: Star[] = [];
    try {

      //mock
      stars.push(peopleMock[0],peopleMock[1]);
      directors.push(peopleMock[2],peopleMock[2]);

      //real data
      //TODO - have pagination like in movies page here
      // stars = await this.starsService.getPagination();
      // directors = await this.directorsService.getPagination
    }
    catch (e) {
      this.toastrService.show('danger', 'Fetching people went wrong.');

    }

    newState = produce(currentState, draft => {
      draft.isFetching = false;
      draft.stars = stars;
      draft.directors = directors;
    })

    setState(newState);
  }

  @Action(PeopleSearchName)
  async peopleSearchName(
    { getState, setState }: StateContext<PeopleStateModel>,
    action: PeopleSearchName) {

    let currentState = getState();

    let newState = produce(currentState, draft => {
      draft.isFetching = true;
    })

    setState(newState);
    currentState = newState;

    let stars: Star[] = [];

    try {
      stars = await this.starsService.getStarsByName(action.personName);
    }
    catch (e){

    }
    newState = produce(currentState, draft => {
      if(stars.length < 0) {
        this.toastrService.show('', 'There are no people to search');
      }
      else {
        let filteredStars = stars.filter((star: Star) => {
          let value = star.name.toLowerCase().includes(action.personName);
          return value;
        });
        if(filteredStars.length < 0) {
          this.toastrService.show('', 'There are no people that contain name ' + action.personName);
        }
        draft.stars = filteredStars;
      }
      draft.isFetching = false;
      draft.isFiltered = true;
    })

    setState(newState);
  }

  @Action(PeopleSearchReset)
  async peopleSearchReset(
    { getState, setState }: StateContext<PeopleStateModel>) {

    let currentState = getState();

    let newState = produce(currentState, draft => {
      draft.isFiltered = false;
      //TODO implement same pagination here
      // draft.stars = this.allPeople;
    })

    setState(newState);
  }

  @Action(PeopleReset)
  async peopleReset(
    { getState, setState }: StateContext<PeopleStateModel>) {
    setState(defaultsState);
  }
}

