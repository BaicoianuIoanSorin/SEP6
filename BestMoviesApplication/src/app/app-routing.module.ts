import { NgModule } from '@angular/core';
import { RouterModule, Routes } from '@angular/router';
import {OverallInformationComponent} from "./overall-information/overall-information.component";
import {NotFoundComponent} from "src/app/not-found/not-found.component";
import {MoviesComponent} from "src/app/information/movies/movies.component";
import {PeopleComponent} from "src/app/information/people/people.component";

const routes: Routes = [
  {
    path: 'overall-information',
    title: 'Overall Information',
    component: OverallInformationComponent,
  },
  {
    path: 'information',
    children: [
      {
        path:'movies',
        title: 'Information Movies',
        component: MoviesComponent,
      },
      {
        path:'people',
        title: 'Information People',
        component: PeopleComponent,
      }
    ]
  },
  { path: '',
    redirectTo: '/overall-information',
    pathMatch: 'full'
  },
  {
    path: '**',
    title: '404 Page not found',
    component: NotFoundComponent
  }
];

@NgModule({
  imports: [RouterModule.forRoot(routes)],
  exports: [RouterModule]
})
export class AppRoutingModule { }
