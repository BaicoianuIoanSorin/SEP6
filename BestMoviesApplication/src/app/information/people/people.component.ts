import {Component, OnDestroy, OnInit} from '@angular/core';
import {NbSearchService} from "@nebular/theme";
import {Select, Store} from "@ngxs/store";
import {Observable} from "rxjs";
import {PeopleSelector} from "src/app/information/people/people.selector";
import {
  PeopleFetchInfoFirstPage,
  PeopleReset,
  PeopleSearchDirectorsReset,
  PeopleSearchStarsByName,
  PeopleSearchStarsReset
} from "src/app/information/people/people.actions";
import {Star} from "src/model/star";
import {Director} from "src/model/director";
import {PeoplePlaceHolderEnum} from "src/app/information/people/constants/constants";
import {Router} from "@angular/router";
import {PeopleType} from "src/app/information/people/people-overview/constants/constants";

@Component({
  selector: 'app-people',
  templateUrl: './people.component.html',
  styleUrls: ['./people.component.scss']
})
export class PeopleComponent implements OnInit, OnDestroy {
  //selectors observable ngxs
  @Select(PeopleSelector.isFetching)
  isFetching$: Observable<boolean>;

  @Select(PeopleSelector.starsAreFiltered)
  starsAreFiltered$: Observable<boolean>;

  @Select(PeopleSelector.starsAreFiltered)
  directorsAreFiltered$: Observable<boolean>;

  @Select(PeopleSelector.stars)
  stars$: Observable<Star[]>;

  @Select(PeopleSelector.directors)
  directors$: Observable<Director[]>;

  starsPlaceholder: string;
  directorsPlaceholder: string;
  alive: boolean = true;

  constructor(private store: Store,
              private router: Router,
              private searchService: NbSearchService) {
  }

  async ngOnInit() {
    this.starsPlaceholder = PeoplePlaceHolderEnum.STARS_PLACEHOLDER;
    this.directorsPlaceholder = PeoplePlaceHolderEnum.DIRECTORS_PLACEHOLDER;

    const actionsInParallel = [
      new PeopleFetchInfoFirstPage(),
    ];
    this.store.dispatch([...actionsInParallel]);
  }

  onSearch(peopleType: PeopleType, event) {
    if (peopleType === PeopleType.STAR) {
      this.starsPlaceholder = event;
      this.store.dispatch(new PeopleSearchStarsByName(event))
    } else if (peopleType === PeopleType.DIRECTOR) {
      this.directorsPlaceholder = event;
      this.store.dispatch(new PeopleSearchStarsByName(event))
    }
  }

  resetSearch(peopleType: PeopleType) {
    if (peopleType === PeopleType.STAR) {
      this.starsPlaceholder = PeoplePlaceHolderEnum.STARS_PLACEHOLDER;
      this.store.dispatch(new PeopleSearchStarsReset());
    } else if (peopleType === PeopleType.DIRECTOR) {
      this.directorsPlaceholder = PeoplePlaceHolderEnum.DIRECTORS_PLACEHOLDER;
      this.store.dispatch(new PeopleSearchDirectorsReset());
    }
  }

  loadNext(peopleType: PeopleType) {
    if (peopleType === PeopleType.STAR) {

    }
    else if(peopleType === PeopleType.DIRECTOR) {

    }
  }

  redirectToOverviewPageFor(peopleType: PeopleType, starId: number) {
    this.router.navigate([`/information/people/${peopleType}/${starId}`]);
  }

  ngOnDestroy() {
    this.alive = false;
    this.store.dispatch(new PeopleReset());
  }

  getPeopleType(): typeof PeopleType {
    return PeopleType;
  }
}
