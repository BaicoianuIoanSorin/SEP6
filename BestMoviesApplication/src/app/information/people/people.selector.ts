import {Selector} from "@ngxs/store";
import {PeopleState, PeopleStateModel} from "src/app/information/people/people.state";

export class PeopleSelector {
  @Selector([PeopleState])
  static isFetching(state: PeopleStateModel) {
    return state.isFetching;
  }

  @Selector([PeopleState])
  static isFiltered(state: PeopleStateModel) {
    return state.starsAreFiltered;
  }

  @Selector([PeopleState])
  static stars(state: PeopleStateModel) {
    return state.stars;
  }

  @Selector([PeopleState])
  static directors(state: PeopleStateModel) {
    return state.directors;
  }
}
