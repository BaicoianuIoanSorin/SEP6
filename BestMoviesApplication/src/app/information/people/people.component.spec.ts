import {NgxsModule, Store} from "@ngxs/store";
import {Spectator} from "@ngneat/spectator";
import {createComponentFactory, SpectatorFactory} from "@ngneat/spectator/jest";
import {RouterTestingModule} from "@angular/router/testing";
import {BrowserModule} from "@angular/platform-browser";
import {AppRoutingModule} from "../../app-routing.module";
import {BrowserAnimationsModule} from "@angular/platform-browser/animations";
import {
  NbActionsModule,
  NbCardModule,
  NbIconModule,
  NbLayoutModule,
  NbMenuModule, NbSearchModule,
  NbSidebarModule,
  NbThemeModule,
  NbToastrModule
} from "@nebular/theme";
import {defaultsState} from "../people/people.state";
import {NbEvaIconsModule} from "@nebular/eva-icons";
import {FormsModule} from "@angular/forms";
import {AgGridModule} from "ag-grid-angular";
import {NEBULAR_MODULES, PROVIDERS, STATES} from "../../app.module";
import {PeopleComponent} from "./people.component";
import {HttpClientModule} from "@angular/common/http";
import {environment} from "../../../environments/environment";

describe('app > information > people > people.component.spec.ts', () => {
  let store: Store;
  let component: PeopleComponent;
  let spectator: Spectator<PeopleComponent>;
  const createComponent: SpectatorFactory<PeopleComponent> =
    createComponentFactory({
      component: PeopleComponent,
      imports: [
        RouterTestingModule,
        BrowserModule,
        HttpClientModule,
        AppRoutingModule,
        BrowserAnimationsModule,
        NgxsModule.forRoot(STATES, {
          developmentMode: !environment.production,
        }),
        ...NEBULAR_MODULES,
      ],
      declarations: [],
      providers: [
        ...PROVIDERS,
        Store,
      ],
    });

  beforeEach(async () => {
    spectator = createComponent({
      detectChanges: false,
    });

    store = spectator.inject(Store);

    store.reset({
      peoplePage: defaultsState,
    })
  });

  it('should create', async () => {
    spectator.detectChanges();
    await spectator.fixture.whenStable();
    spectator.detectChanges();
    await spectator.fixture.whenRenderingDone();

    expect(spectator.component).toBeTruthy();
  });

  it('should match snapshot', async () => {
    spectator.detectChanges();
    await spectator.fixture.whenStable();
    spectator.detectChanges();
    await spectator.fixture.whenRenderingDone();

    // assert
    expect(spectator.fixture).toMatchSnapshot();
  });
});
