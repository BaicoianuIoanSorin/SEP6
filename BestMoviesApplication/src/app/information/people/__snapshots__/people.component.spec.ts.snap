// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app > information > people > people.component.spec.ts should match snapshot 1`] = `
<app-people
  __isFetching$__selector={[Function Observable]}
  alive={[Function Boolean]}
  directorsPlaceholder={[Function String]}
  router={[Function Router]}
  searchService={[Function NbSearchService]}
  starsPlaceholder={[Function String]}
  store={[Function Store]}
>
  <nb-spinner
    class="spinner size-medium status-basic"
    status="basic"
  >
    <span
      class="spin-circle"
    />
    <span
      class="message"
    >
      Loading...
    </span>
  </nb-spinner>
</app-people>
`;
